### PR TITLE
HZN-1185: Add support for OSGi based Telemetryd adapters

### PR DESCRIPTION
**Pull request by cgorantla**
HZN-1185: Add support for OSGi based Telemetryd adapters

Add jti adapter osgi feature. 
Move OSGiScript related files to common project

* JIRA: http://issues.opennms.org/browse/hzn-1185

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

### CODE REVIEW

File: checkstyle/src/main/resources/license-suppressions.xml [hunk 1]
```diff
@@ -13,6 +13,7 @@
     <suppress files="org/opennms/features/topology/plugins/topo/graphml/internal/scripting/.*" checks=".*"/>
     <suppress files="org/opennms/poller/remote/Mac.*" checks=".*"/>
     <suppress files="org/opennms/container/web/.*" checks=".*"/>
+    <suppress files="org/opennms/features/osgi/script/.*" checks=".*"/>
```
Comment (j-white) [inline checkstyle/src/main/resources/license-suppressions.xml]: We can remove the previous entry for these since we changed the package name.

File: container/features/src/main/resources/features.xml [hunk 2]
```diff
@@ -1077,4 +1077,12 @@
       <bundle>mvn:org.opennms.features.poller/org.opennms.features.poller.shell/${project.version}</bundle>
     </feature>
 
+    <feature name="opennms-telemetry-jti" description="OpenNMS :: Telemetry :: JTI" version="${project.version}">
+      <feature>opennms-collection-api</feature>
+      <feature version="${guavaVersion}">guava</feature>
+      <bundle>mvn:com.google.protobuf/protobuf-java/${protobufVersion}</bundle>
```
Comment (j-white) [inline container/features/src/main/resources/features.xml]: Do we actually depend on protobuf?

File: container/karaf/src/main/filtered-resources/etc/custom.properties [hunk 3]
```diff
@@ -795,5 +795,10 @@ org.osgi.framework.system.packages.extra=org.apache.karaf.branding,\
         org.opennms.web.svclayer.model;version=${opennms.osgi.version},\
         org.opennms.web.svclayer.support;version=${opennms.osgi.version},\
         org.opennms.netmgt.icmp;version=${opennms.osgi.version},\
-        org.opennms.netmgt.icmp.proxy;version=${opennms.osgi.version}
+        org.opennms.netmgt.icmp.proxy;version=${opennms.osgi.version},\
+        org.opennms.netmgt.filter.api;version=${opennms.osgi.version},\
```
Comment (j-white) [inline container/karaf/src/main/filtered-resources/etc/custom.properties]: These should be in alphabetical order instead of just appending them to the end.

We could also fix up the the `org.opennms.netmgt.icmp` and `org.opennms.netmgt.icmp.proxy` entries while were at it.

File: features/telemetry/adapters/api/src/main/java/org/opennms/netmgt/telemetry/adapters/api/Adapter.java [hunk 4]
```diff
@@ -38,7 +38,7 @@
  *
  * @author jwhite
  */
-public interface Adapter {
+public interface Adapter{
```
Comment (j-white) [inline features/telemetry/adapters/api/src/main/java/org/opennms/netmgt/telemetry/adapters/api/Adapter.java]: Unnecessary change.

File: container/features/src/main/resources/features.xml [hunk 5]
```diff
@@ -1077,4 +1077,12 @@
...
+      <bundle>mvn:org.opennms.features.telemetry.adapters/org.opennms.features.telemetry.adapters.registry/${project.version}</bundle>
```
Comment (j-white) [inline container/features/src/main/resources/features.xml]: I don't think this feature depends on the registry either.

File: features/telemetry/adapters/jti/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		http://www.osgi.org/xmlns/blueprint/v1.0.0 
+		http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd
+		http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0
+		http://aries.apache.org/schemas/blueprint-cm/blueprint-cm-1.1.0.xsd
+		http://aries.apache.org/blueprint/xmlns/blueprint-ext/v1.1.0
+		http://aries.apache.org/schemas/blueprint-ext/blueprint-ext-1.1.xsd
+">
+
+	<bean id="jtiFactory" class="org.opennms.netmgt.telemetry.adapters.jti.JtiAdapterFactory">
+	    <argument ref="blueprintBundleContext" />
```
Comment (j-white) [inline features/telemetry/adapters/jti/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Nit: This one isn't lined up with the others :)

File: features/telemetry/adapters/jti/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+			<entry key="type" value="org.opennms.netmgt.telemetry.adapters.jti.JtiGpbAdapter" />
+		</service-properties>
+	</service>
+
+	<reference id="collectionAgentFactory" interface="org.opennms.netmgt.collection.api.CollectionAgentFactory"/>
+	<reference id="interfaceToNodeCache" interface="org.opennms.netmgt.dao.api.InterfaceToNodeCache"/>
+	<reference id="nodeDao" interface="org.opennms.netmgt.dao.api.NodeDao"/>
+	<reference id="filterDao" interface="org.opennms.netmgt.filter.api.FilterDao"/>
+	<reference id="transactionTemplate" interface="org.springframework.transaction.support.TransactionOperations"/>
+    <reference id="persisterFactory" interface="org.opennms.netmgt.collection.api.PersisterFactory"/>
```
Comment (j-white) [inline features/telemetry/adapters/jti/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Nit: This one either :)

File: features/telemetry/adapters/registry/pom.xml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+      <groupId>org.opennms.features.telemetry.adapters</groupId>
+      <artifactId>org.opennms.features.telemetry.adapters.api</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.opennms.features.telemetry.adapters</groupId>
+      <artifactId>org.opennms.features.telemetry.adapters.factory</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+	<dependency>
```
Comment (j-white) [inline features/telemetry/adapters/registry/pom.xml]: Nit: White space is off here.

File: features/telemetry/daemon/src/main/java/org/opennms/netmgt/telemetry/daemon/TelemetryMessageConsumer.java [hunk 9]
```diff
@@ -99,34 +106,59 @@ public void handleMessage(TelemetryProtos.TelemetryMessageLog messageLog) {
     }
 
     private Adapter buildAdapter(org.opennms.netmgt.telemetry.config.model.Adapter adapterDef) throws Exception {
```
Comment (j-white) [inline features/telemetry/daemon/src/main/java/org/opennms/netmgt/telemetry/daemon/TelemetryMessageConsumer.java]: I think we should move all the logic about constructing an adapter into the registry instead of having it here.

File: features/telemetry/daemon/src/main/java/org/opennms/netmgt/telemetry/daemon/TelemetryMessageConsumer.java [hunk 10]
```diff
@@ -99,34 +106,59 @@ public void handleMessage(TelemetryProtos.TelemetryMessageLog messageLog) {
...
-        } catch (Exception e) {
-            throw new RuntimeException(String.format("Failed to instantiate adapter with class name '%s'.",
-                    adapterDef.getClassName(), e));
-        }
 
-        // Cast
-        if (!(adapterInstance instanceof Adapter)) {
-            throw new IllegalArgumentException(String.format("%s must implement %s", adapterDef.getClassName(), Adapter.class.getCanonicalName()));
+        AdapterFactory adapterFactory = null;
+        while (ManagementFactory.getRuntimeMXBean().getUptime() < GRACE_PERIOD_MS) {
```
Comment (j-white) [inline features/telemetry/daemon/src/main/java/org/opennms/netmgt/telemetry/daemon/TelemetryMessageConsumer.java]: If the daemon is reloaded after the system has been up, we won't even attempt a lookup in the registry.

File: container/features/src/main/resources/features.xml (hunk 2 above)
Comment (cgorantla) [reply container/features/src/main/resources/features.xml]: We do depend on protobuf in jti adapter.

File: features/telemetry/daemon/src/main/java/org/opennms/netmgt/telemetry/daemon/TelemetryMessageConsumer.java (hunk 9 above)
Comment (cgorantla) [reply features/telemetry/daemon/src/main/java/org/opennms/netmgt/telemetry/daemon/TelemetryMessageConsumer.java]: Took care of all the review comments.

File: container/features/src/main/resources/features.xml (hunk 2 above)
Comment (j-white) [reply container/features/src/main/resources/features.xml]: Duh, of course. I was confusing this with the protobuf messages used between OpenNMS and Minion.

**Comment by j-white** [comment]
Any input on this @mvrueden?  We should be able to migrate the Netflow 5 adapter in the `features/drift` branch to OSGi with these changes.

File: container/features/src/main/resources/features.xml [hunk 11]
```diff
@@ -1077,4 +1077,11 @@
       <bundle>mvn:org.opennms.features.poller/org.opennms.features.poller.shell/${project.version}</bundle>
     </feature>
 
+    <feature name="opennms-telemetry-jti" description="OpenNMS :: Telemetry :: JTI" version="${project.version}">
```
Comment (mvrueden) [inline container/features/src/main/resources/features.xml]: Usually we let the `maven-features-plugin` take care of the feature generation. Is there a particular reason why the feature is defined manually?

File: features/telemetry/adapters/collection/src/main/java/org/opennms/netmgt/telemetry/adapters/collection/ScriptedCollectionSetBuilder.java [hunk 12]
```diff
@@ -56,13 +59,19 @@
...
 
         final String ext = Files.getFileExtension(script.getAbsolutePath());
-        final ScriptEngineManager manager = new ScriptEngineManager();
+
+        ScriptEngineManager manager;
+        if (bundleContext != null) {
+            manager = new OSGiScriptEngineManager(bundleContext);
+        } else {
+            manager = new ScriptEngineManager();
+        }
```
Comment (mvrueden) [inline features/telemetry/adapters/collection/src/main/java/org/opennms/netmgt/telemetry/adapters/collection/ScriptedCollectionSetBuilder.java]: Can we move this code to the constructor and have something like:

```java
public ScriptedcollectionSetBuilder(File script) {
    this(script, new ScriptEngineManager())
}

public ScriptedCollectionSetBuilder(File script, BundleContext context) {
    this(script, new OSGiScriptEngineManager(context);
}

private ScriptCollectionSetBuilder(File script, ScriptEngineManager manager) {
     ...
}
```

File: features/telemetry/adapters/registry/pom.xml [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    <groupId>org.opennms.features.telemetry</groupId>
+    <artifactId>org.opennms.features.telemetry.adapters</artifactId>
+    <version>22.0.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>org.opennms.features.telemetry.adapters</groupId>
+  <artifactId>org.opennms.features.telemetry.adapters.registry</artifactId>
+  <name>OpenNMS :: Features :: Telemetry :: Adapters :: Registry</name>
+  <packaging>bundle</packaging>
+  <build>
```
Comment (mvrueden) [inline features/telemetry/adapters/registry/pom.xml]: The `maven-bundle-plugin` setting is duplicated a lot. We may want to define it in a parent module instead.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.graphml/pom.xml [hunk 14]
```diff
@@ -103,6 +103,11 @@
       <artifactId>org.osgi.core</artifactId>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>org.opennms.features</groupId>
+      <artifactId>org.opennms.features.osgi-jsr223</artifactId>
+      <version>${project.version}</version>
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.graphml/pom.xml]: scope should be provided

File: container/features/src/main/resources/features.xml (hunk 11 above)
Comment (j-white) [reply container/features/src/main/resources/features.xml]: Personally I like to avoid using our `maven-features-plugin`: it adds unnecessary bundles, and doesn't reference other features that already exist.

Comment (mvrueden) [reply container/features/src/main/resources/features.xml]: If it is setup correctly it will not add unnecessary bundles. It only adds transitive dependencies if the scope is not set to provided. Also features are supported. Only thing I see which makes it hard, is that you cannot define a version on a feature.

Comment (cgorantla) [reply container/features/src/main/resources/features.xml]: I always had features defined here except for some of the minion features.  Unless there is specific advantage, I will keep it here.

File: features/telemetry/adapters/registry/pom.xml (hunk 13 above)
Comment (cgorantla) [reply features/telemetry/adapters/registry/pom.xml]: parent module is adapters which just includes other modules. There may be other adapters which may not be osgi bundles. I will keep  it that way.  Also good to have bundle info in it's own pom.xml I feel.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.graphml/pom.xml (hunk 14 above)
Comment (cgorantla) [reply features/topology-map/plugins/org.opennms.features.topology.plugins.topo.graphml/pom.xml]: I updated scope here. But in case of adapters-collection module,  I have it as default scope. We need to have it work on Spring as well if feature doesn't exist, so OSGiScriptEngine related files needs to be in default classpath.

File: features/telemetry/adapters/registry/pom.xml (hunk 13 above)
Comment (mvrueden) [reply features/telemetry/adapters/registry/pom.xml]: The definition here, only extends maven, which allows defining bundles. It does not make all modules bundles. The type is defined by the `<type>bundle</type>` statement. Therefore I am still for moving it to a parent module.

File: container/features/src/main/resources/features.xml (hunk 11 above)
Comment (mvrueden) [reply container/features/src/main/resources/features.xml]: The specific advantage is, that the `maven-features-plugin` will add/remove a bundle to the definition every time the pom of the module changes.
